File: backend/src/models/ContractTemplate.entity.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToMany,
  ManyToOne,
  JoinColumn,
  JoinTable,
} from "typeorm";
import { User } from "./User.entity";
import { Partner } from "./Partner.entity";

@Entity("contract_templates")
export class ContractTemplate {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: "varchar", length: 255 })
  title: string;

  @Column({ type: "text" })
  content: string;

  @Column({ type: "varchar", default: "active" })
  status: string;

  @Column({ type: "varchar", default: "richtext" })
  template_type: string;

  @Column({ type: "text", nullable: true })
  pdf_url: string;

  @Column({ type: "jsonb", default: "[]" })
  field_positions: Array<{
    name: string;
    x: number;
    y: number;
    width: number;
    height: number;
    page: number;
    type?: string;
  }>;

  @Column({ type: "json", nullable: true })
  dropdown_fields: {
    [fieldName: string]: {
      label: string;
      options: Array<{
        label: string;
        value: string;
      }>;
      required?: boolean;
      placeholder?: string;
    };
  };

  @ManyToMany(() => User)
  @JoinTable({
    name: "contract_template_sales_reps",
    joinColumn: { name: "contract_template_id", referencedColumnName: "id" },
    inverseJoinColumn: { name: "sales_rep_id", referencedColumnName: "user_id" },
  })
  assigned_sales_reps: User[];

  @Column({ type: "int", nullable: true })
  partner_id: number;

  @ManyToOne(() => Partner, { onDelete: "SET NULL" })
  @JoinColumn({ name: "partner_id" })
  partner: Partner;

  @CreateDateColumn()
  created_at: Date;

  @UpdateDateColumn()
  updated_at: Date;
}

File: salesman-app/lib/src/features/visits/domain/entities/contract_template_entity.dart
import 'package:equatable/equatable.dart';

class DropdownOption extends Equatable {
  final String label;
  final String value;

  const DropdownOption({
    required this.label,
    required this.value,
  });

  @override
  List<Object?> get props => [label, value];
}

class DropdownField extends Equatable {
  final String label;
  final List<DropdownOption> options;
  final bool required;
  final String? placeholder;

  const DropdownField({
    required this.label,
    required this.options,
    this.required = false,
    this.placeholder,
  });

  @override
  List<Object?> get props => [label, options, required, placeholder];
}

class ContractTemplateEntity extends Equatable {
  final int id;
  final String templateString;
  final String title;
  final Map<String, DropdownField>? dropdownFields;
  final String templateType;
  final String? pdfUrl;
  final List<Map<String, dynamic>>? fieldPositions;

  const ContractTemplateEntity({
    required this.id,
    required this.title,
    required this.templateString,
    this.dropdownFields,
    this.templateType = 'richtext',
    this.pdfUrl,
    this.fieldPositions,
  });

  bool get isPdfUpload => templateType == 'pdf_upload';

  @override
  List<Object?> get props => [id, title, templateString, dropdownFields, templateType, pdfUrl, fieldPositions];
}

class ContractTemplateListEntity extends Equatable {
  final List<ContractTemplateEntity> contracts;

  const ContractTemplateListEntity({required this.contracts});

  @override
  List<Object?> get props => [contracts];
}
